File: 02.xtrem-t.d.d/src/test/java/money_problem/domain/PortfolioTest.java
package money_problem.domain;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

public class PortfolioTest {
    private Bank bank;

    @BeforeEach
    void setUp() {
        bank = Bank.withExchangeRate(Currency.USD, Currency.USD, 1.0)
                .addExchangeRate(Currency.USD, Currency.KRW, 1100)
                .addExchangeRate(Currency.EUR, Currency.USD, 1.2);
    }

    @Test
    void testPortfolioCanEvaluateWithinTheSameCurrency() {
        Portfolio portfolio = new Portfolio()
                .addMoney(new Money(5.0, Currency.USD))
                .addMoney(new Money(10.0, Currency.USD));

        Assertions.assertThat(portfolio.evaluate(bank, Currency.USD)).isEqualTo("15.0 USD");
    }

    @Test
    @DisplayName("5 USD + 10 EUR = 17 USD")
    void testEvaluateMultiCurrencies() {
        Portfolio portfolio = new Portfolio()
                .addMoney(new Money(5.0, Currency.USD))
                .addMoney(new Money(10.0, Currency.EUR));

        Assertions.assertThat(portfolio.evaluate(bank, Currency.USD)).isEqualTo("17.0 USD");
    }

    @Test
    @DisplayName("1 USD + 1100 KRW = 2200 KRW")
    void testEvaluateMultiCurrenciesWithKoreanWon() {
        Portfolio portfolio = new Portfolio()
                .addMoney(new Money(1.0, Currency.USD))
                .addMoney(new Money(1100.0, Currency.KRW));

        Assertions.assertThat(portfolio.evaluate(bank, Currency.KRW)).isEqualTo("2200.0 KRW");
    }
}

File: 02.xtrem-t.d.d/src/test/kotlin/money_problem/core/PortfolioTest.kt
package money_problem.core

import io.kotest.assertions.arrow.core.shouldBeLeft
import io.kotest.assertions.arrow.core.shouldBeRight
import io.kotest.core.spec.style.StringSpec
import money_problem.core.Currency.*

class PortfolioTest : StringSpec({
    val bank = Bank.withExchangeRate(USD, USD, 1.0)
        .addExchangeRate(USD, KRW, 1100.0)
        .addExchangeRate(EUR, USD, 1.2)

    "5 USD + 10 USD = 15 USD" {
        portfolioWith(
            5.0.dollars(),
            10.0.dollars()
        ).evaluate(bank, USD) shouldBeRight 15.0.dollars()
    }

    "5 USD + 10 EUR = 17 USD" {
        portfolioWith(
            5.0.dollars(),
            10.0.euros()
        ).evaluate(bank, USD) shouldBeRight 17.0.dollars()
    }

    "1 USD + 1100 KRW = 2220 KRW" {
        portfolioWith(
            1.0.dollars(),
            1100.0.koreanWons()
        ).evaluate(bank, KRW) shouldBeRight 2200.0.koreanWons()
    }

    "5 USD + 10 EUR + 4 EUR = 21.8 USD" {
        portfolioWith(
            5.0.dollars(),
            10.0.euros(),
            4.0.euros()
        ).evaluate(bank, USD) shouldBeRight 21.8.dollars()
    }

    "Return a failure result in case of missing exchange rates" {
        val emptyBank = Bank.withExchangeRate(EUR, USD, 1.2)

        portfolioWith(
            1.0.dollars(),
            1.0.euros(),
            1.0.koreanWons()
        ).evaluate(emptyBank, KRW) shouldBeLeft MissingExchangeRatesError(
            listOf(
                MissingExchangeRateError(USD, KRW),
                MissingExchangeRateError(EUR, KRW)
            )
        )
    }

})

fun portfolioWith(vararg moneys: Money): Portfolio = moneys.fold(Portfolio(), Portfolio::addMoney)
